content/posts/pandastutor.md

```diff
@@ -0,0 +1,323 @@
+  const result = pyodide.runPython('pandas_tutor.pyodide_main.run_code_from_js()');
+  window.globalUserCode = undefined; // null this out afterward to prevent leakage
+  return result;
+}
+```
+
+We couldn't figure out a more elegant way to pass arguments from
+JavaScript to Python functions, so we used a `globalUserCode` variable
+as a hack that seems to work for now.
+
+
+### Step 3: import the Pandas Tutor module in JavaScript
+
+Our final setup step was adding a line of code to our frontend
+JavaScript to import the Pandas Tutor module:
+
+```js
+pyodide.runPython('import pandas_tutor.pyodide_main');
+```
+
+This is a synchronous call that imports `pandas_tutor` and all
+dependencies when the web app first loads. This line takes 3-5 seconds
+to finish since dependencies (including pandas) must be loaded.
+Unfortunately, browser caching can't speed things up here since even if
+all the files are cached locally, Pyodide still needs to run the import
+and initialization code at page load time.
+
+One feature addition that might speed up this process is if Pyodide
+could somehow *serialize a binary memory snapshot* taken right after
+everything gets initialized. Then the user's browser can directly load
+that snapshot on the first page visit and cache it for future visits.
+(But maybe that snapshot would be too big to download efficiently, so
+nevermind!)
+
+
+### Bonus: other fine-tuning and ongoing work
+
+- [Graceful
+  degradation](https://developer.mozilla.org/en-US/docs/Glossary/Graceful_degradation):
+  To continue supporting older browsers, our code detects if there are
+  any problems loading the Pyodide version of Pandas Tutor and, if so,
+  automatically reverts back to the original server-side version. Also,
+  [loadPyodide](https://pyodide.org/en/stable/usage/api/js-api.html#globalThis.loadPyodide)
+  crashes on some older mobile devices (presumably with [limited
+  memory](https://github.com/pyodide/pyodide/issues/533)?) and we can't
+  seem to catch the exception to recover; so we disable it by default on
+  mobile and add a toggle to let the user explicitly enable it. (This is
+  also nice since it prevents the page from preemptively downloading
+  large amounts of data on users' mobile phone plans.)
+- Auto-importer: We want users to be able to import any package when
+  writing their code in Pandas Tutor. But Pyodide's
+  [loadPackagesFromImports](https://pyodide.org/en/stable/usage/api/js-api.html#pyodide.loadPackagesFromImports)
+  works only on packages that come with Pyodide. To give users more
+  flexibility, we wrote an auto-importer that automatically calls
+  `micropip.install()` whenever the user's code tries to `import` an
+  unknown package. These package wheels get downloaded from PyPI
+  on-demand and cached in the browser. (Note that this works only if the
+  package name on PyPI matches the name in the `import` statement, but
+  this is often true.)
```

Comment:
Cool. Just for information that is exactly a reason why we don't auto-install things from PyPi. The other being typos in imports etc. We want Pyodide APIs to be 100% reliable, so that downstream applications can build things on top of them as you are doing here.

Reply:
yep, totally makes sense!